packages/upload-api/src/blob/add.js

```diff
@@ -0,0 +1,216 @@
+        }
+      }
+
+      // We derive principal from the content multihash to be an audience
+      // of the `http/put` invocation. That way anyone with blob content
+      // could perform the invocation and issue receipt by deriving same
+      // principal
+      const putSubject = await ed25519.derive(blob.content.slice(0, 32))
+      const facts = [
+        {
+          keys: putSubject.toArchive(),
+        },
+      ]
+
+      // Create web3.storage/blob/* invocations
+      const blobAllocate = W3sBlob.allocate.invoke({
+        issuer: id,
+        audience: id,
+        with: id.did(),
+        nb: {
+          blob,
+          cause: invocation.link(),
+          space,
+        },
+        expiration: Infinity,
+      })
+      const blobAccept = W3sBlob.accept.invoke({
+        issuer: id,
+        audience: id,
+        with: id.toDIDKey(),
+        nb: {
+          blob,
+          exp: Number.MAX_SAFE_INTEGER,
+        },
+        expiration: Infinity,
+      })
+      const [allocatefx, acceptfx] = await Promise.all([
+        blobAllocate.delegate(),
+        blobAccept.delegate(),
+      ])
+
+      // Get receipt for `blob/allocate` if available, or schedule invocation if not
+      const allocatedGetRes = await allocationsStorage.get(space, blob.content)
+      let blobAllocateReceipt
+      let blobAllocateOutAddress
+      // If already allocated, just get the allocate receipt
+      // and the addresses if still pending to receive blob
+      if (allocatedGetRes.ok) {
+        const receiptGet = await context.receiptsStorage.get(allocatefx.link())
+        if (receiptGet.error) {
+          return receiptGet
+        }
+        blobAllocateReceipt = receiptGet.ok
+
+        // Check if despite allocated, the blob is still not stored
+        const blobHasRes = await context.blobsStorage.has(blob.content)
+        if (blobHasRes.error) {
+          return blobHasRes
+        } else if (!blobHasRes.ok) {
+          // @ts-expect-error receipt type is unknown
+          blobAllocateOutAddress = blobAllocateReceipt.out.ok.address
+        }
+      }
+      // if not already allocated, schedule `blob/allocate`
+      else {
+        // Execute allocate invocation
+        const allocateRes = await blobAllocate.execute(getServiceConnection())
+        if (allocateRes.out.error) {
+          return {
+            error: allocateRes.out.error,
+          }
+        }
+        // If this is a new allocation, `http/put` effect should be returned with address
+        blobAllocateOutAddress = allocateRes.out.ok.address
+        blobAllocateReceipt = allocateRes
+      }
+
+      // Create `blob/allocate` receipt invocation to inline as effect
+      const message = await Message.build({ receipts: [blobAllocateReceipt] })
+      const messageCar = await CAR.outbound.encode(message)
+      const bytes = new Uint8Array(messageCar.body)
+      const messageLink = await CAR.codec.link(bytes)
```

Comment:
I already raised this somewhere else, but can't find the comment now. We can only attach blocks that have a `link` in the nb field. Doing what you suggest below with iterating through receipt blocks leads to 
```
HandlerExecutionError: service handler {can: "blob/add"} error: given block with bafyreieaipb7idctqbfmgzpb4mfjy7msgmncng7qieiqssbshv5v5z7hci is not an attached link
```

Moreover, the message allows us to easily handle breaking changes. So, while I started to use it as a workaround for the above, I now actually give it the value of message version. What do you think?

Reply:
> Moreover, the message allows us to easily handle breaking changes

I think having two version of message one outer and one inner is probably going to be more pain than help, however that is just an intuition. I'm not opposed to it, but would prefer to avoid if we can.

Per our call today, I'm going to see if I can lift the limitations in ucanto to make this work